eif_utils/src/lib.rs

```diff
@@ -519,15 +669,71 @@ impl<T: Digest + Debug + Write + Clone> EifBuilder<T> {
         if self.sign_info.is_some() {
             self.generate_eif_signature(&measurements);
         }
+        self.save_metadata().expect("Failed to save metadata");
         self.compute_crc();
         self.write_header(output_file);
         self.write_kernel(output_file);
         self.write_cmdline(output_file);
         self.write_ramdisks(output_file);
         self.write_signature(output_file);
+        self.write_metadata(output_file);
         measurements
     }
 
+    pub fn get_generated_meta(&mut self) -> BTreeMap<String, String> {
+        let mut meta = BTreeMap::new();
+
+        let time = SystemTime::now();
+        let time: DateTime<Utc> = time.into();
+        meta.insert("BuildTime".to_string(), time.to_rfc3339());
+
+        let blobs_path =
+            std::env::var("NITRO_CLI_BLOBS").unwrap_or_else(|_| DEFAULT_BLOBS_PATH.to_string());
+        // Read aws-nitro-enclaves-cli.spec file and parse the lines that contain:
+        // Name:      aws-nitro-enclaves-cli
+        // Version:   1.0.12
+        let specs_path =
+            std::env::var("NITRO_CLI_SPECS").unwrap_or_else(|_| DEFAULT_SPECS_PATH.to_string());
+        let specs_file = File::open(format!("{}/aws-nitro-enclaves-cli.spec", specs_path))
+            .expect("Failed to open NitroCLI specs file.");
+        let spec_lines = BufReader::new(specs_file).lines();
+        let mut tool_name = String::new();
+        let mut tool_version = String::new();
+
+        for line in spec_lines {
+            let line = line.expect("Failed to extract line from specs");
+            let clean_line = line.replace(' ', "");
+            let split_line: Vec<&str> = clean_line.split(':').collect();
+
+            if line.contains("Name:") {
+                tool_name = split_line[1].to_string();
+            } else if line.contains("Version:") {
+                tool_version = split_line[1].to_string();
+                break;
+            }
+        }
+
+        meta.insert("BuildTool".to_string(), tool_name);
+        meta.insert("BuildToolVersion".to_string(), tool_version);
+
+        let config_file = File::open(format!("{}/{}", blobs_path, CONFIG_FILE))
+            .expect("Failed to open kernel image config file.");
+        let os_string: String = BufReader::new(config_file)
+            .lines()
+            .nth(2)
+            .unwrap()
+            .expect("Failed to read kernel config file.");
+
+        // Extract OS and version from line format:
+        // ' # Linux/x86_64 4.14.177-104.253.amzn2.x86_64 Kernel Configuration '
+        let sep: Vec<char> = vec![' ', '/', '-'];
+        let os_words: Vec<&str> = os_string.split(&sep[..]).collect();
+        meta.insert("OperatingSystem".to_string(), os_words[1].to_string());
```

Comment:
Same observation as above: Check first that indices 1 and 3 are valid and consider using defaults if the tokenized string doesn't contain valid words at indices 1 or 3.

Reply:
Added the suggested checks.